Defensive cleanup

djih opened the pull request.

djih commented on Makefile:
```diff
@@ -86,6 +87,9 @@ build: $(TESTS) $(OUT) $(SNIPPET_OUT) $(SEGMENT_SNIPPET_OUT) README.md
 	@$(DUO) --development test/tests.js > build/tests.js
 	@$(DUO) --development test/snippet-tests.js > build/snippet-tests.js
 
+docs:
```
I considered adding this to the default make command, but the html files are timestamped, so they would be modified with every make

djih commented on README.md:
````diff
@@ -224,18 +226,25 @@ amplitude.init('YOUR_API_KEY_HERE', null, {
 ```
 
-| option | description | default |
+| option | type | description | default |
````
Added a type column

djih commented on src/amplitude.js:
```diff
@@ -13,83 +14,53 @@ var UUID = require('./uuid');
-Amplitude.prototype._sessionId = null;
-Amplitude.prototype._newSession = false;
-Amplitude.prototype._updateScheduled = false;
+  // event meta data
```
We were doing something weird with declaring instance variables in both the constructor and on the prototype. We should just stick everything in the constructor

paladin8 replied:
This makes it so you can't access those fields without constructing an instance, right? Is it possible that someone is using the prototype fields right now (probably by accident)?

djih commented on src/amplitude.js:
```diff
@@ -13,83 +14,53 @@ var UUID = require('./uuid');
  */
-Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+Amplitude.prototype.init = function init(apiKey, opt_userId, opt_config, opt_callback) {
+  if (type(apiKey) !== 'string' || utils.isEmptyString(apiKey)) {
```
Aggressively validate input. This includes checking the type, as well as for empty strings

ManThursday commented on documentation/amplitude.js.html:
```diff
@@ -0,0 +1,1062 @@
+ * @param {string} eventType - name of event
+ * @param {object} eventProperties - (optional) an object with string keys and values for the event properties.
+ * @param {function} opt_callback - (optional) a callback function to run after the event is logged.
+ * Note: the server response code and response body from the identify event upload are passed to the callback function.
```
"identify" copy/paste

djih commented on src/amplitude.js:
```diff
@@ -99,10 +70,11 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
-    this.options.userId = (opt_userId !== undefined && opt_userId !== null && opt_userId) || this.options.userId || null;
+    // load deviceId and userId from input, or try to fetch existing value from cookie
+    this.options.deviceId = (type(opt_config) === 'object' && type(opt_config.deviceId) === 'string' &&
+        !utils.isEmptyString(opt_config.deviceId) && opt_config.deviceId) || this.options.deviceId || UUID() + 'R';
```
Appending a 'R' to generated deviceIds to be consistent with iOS and Android SDKs

ManThursday commented on src/amplitude.js:
```diff
@@ -362,53 +480,49 @@ Amplitude.prototype._getReferringDomain = function(referrer) {
+/**
+ * Saves unsent events and identifies to localStorage. JSON stringifies event queues before saving.
+ * Note: this is called automatically every time events are logged, unless you explicitly set option saveEvents to false.
+ * @public
```
I was initially a bit confused by what this meant.  Maybe an example use case would help here.

djih commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
   for (var i = 0; i < this._q.length; i++) {
     var fn = this[this._q[i][0]];
-    if (fn && type(fn) === 'function') {
+    if (type(fn) === 'function') {
```
can skip the null check since type(null) will return 'null' which fails the === 'function' test

djih commented:
Refactoring a bunch of stuff, adding lots of sanity checks and more type-checking around user-input, handle more failure cases gracefully.

1. Use Jsdoc to generate [documentation](https://rawgit.com/amplitude/Amplitude-Javascript/defensive_cleanup/documentation/Amplitude.html) around our SDK classes. This PR looks really big, but 90% of it is just new files (CSS, font, html files) to support the static docs. Also had to update a lot of comments through the SDK to generate the docs.

2. For the Readme section on config inputs, I added a type column to specify the expected input type. The SDK has a helper function that parses the config input object, validates the keys and input types against the DEFAULT config object.

3. Refactored a lot of static strings into Constants.js

4. We can simplify the input validation by leveraging the type library. Before we would do something like this: `if (var === null || type(var) !== 'string')`, but type(null) returns 'null', so you can simply just do a type(var) !== 'string' and that will also check for nulls and undefined

5. Following 4, we make the SDK more defensive by aggressively checking the input, for null, for types, for valid values

6. If we are using a generated deviceId, append an 'R' to the end, to make JS SDK consistent with our iOS and android SDKs

7. Since we validate eventProperties on init, we might as well validate userProperties.

djih commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
-      // utils.log(e);
-    }
+      var events = JSON.parse(savedUnsentEventsString);
+      if (type(events) === 'array') { // handle case where JSON dumping of unsent events is corrupted
```
It's supposed to deserialize into an array object, but if client's have broken JSON libraries, it deserializes into a string, which breaks the SDK

djih commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+ * @private
+ * @deprecated
+ */
+Amplitude.prototype._getFromStorage = function _getFromStorage(storage, key) {
```
Plan to remove _getFromStorage and setInStorage helper functions in v3.0.0

djih replied:
Scratch the above comment, I forgot it's used in multiple apps. The important piece is missing from the code, namely the part where we append the namespace identifier to the key

djih commented on src/amplitude.js:
```diff
@@ -362,53 +480,49 @@ Amplitude.prototype._getReferringDomain = function(referrer) {
-  } catch (e) {
-    // utils.log(e);
-  }
+    localStorage.setItem(this.options.unsentKey, JSON.stringify(this._unsentEvents));
```
Split this up into 2 try blocks. We don't want one load failing to block the other load

djih commented on src/amplitude.js:
```diff
@@ -419,13 +533,18 @@ Amplitude.prototype.setDomain = function(domain) {
+ * @param {string} userId - identifier to set. Can be null.
+ * @example amplitude.setUserId('joe@gmail.com');
+ */
+Amplitude.prototype.setUserId = function setUserId(userId) {
```
We don't check for empty string here since we allow null userIds

djih commented on src/amplitude.js:
```diff
@@ -468,8 +599,16 @@ Amplitude.prototype.setDeviceId = function(deviceId) {
+ * Sets user properties for the current user.
+ * @public
+ * @param {object} - object with string keys and values for the user properties to set.
+ * @param {boolean} - DEPRECATED opt_replace: in earlier versions of the JS SDK the user properties object was kept in
```
A lot of people have asked about where the `replace` flag went, so I added a little explanation

djih commented on src/amplitude.js:
```diff
@@ -787,14 +947,23 @@ Amplitude.prototype._mergeEventsAndIdentifys = function(numEvents) {
+
+    // case 0: no events or identifys left
+    // note this should not happen, this means we have less events and identifys than expected
+    if (noEvents && noIdentifys) {
```
This is the case that crashed on Android SDK. It should in theory never happen, but gracefully handle if it does

djih commented:
@paladin8 you can skip everything in the documentation folder

paladin8 commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+  // verifies config value is defined, is the correct type, and some additional value verification
+  var parseValidateLoad = function parseValidateLoad(key, expectedType) {
+    var input_value = config[key];
+    if(input_value === undefined || !utils.validateInput(input_value, key + ' option', expectedType)) {
```
Space after if.

paladin8 commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+
+  // verifies config value is defined, is the correct type, and some additional value verification
+  var parseValidateLoad = function parseValidateLoad(key, expectedType) {
+    var input_value = config[key];
```
inputValue?

djih replied:
hah yea I had just switched over from python land

paladin8 commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+    return;
   }
+
+  // verifies config value is defined, is the correct type, and some additional value verification
```
Validate seems more appropriate than verify.

paladin8 commented on src/amplitude.js:
```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+    } else {
+      options[key] = (expectedType === 'string' && !utils.isEmptyString(input_value) && input_value) ||
+                     (expectedType === 'number' && config[key] > 0 && input_value) ||
+                     options[key];
```
I think it'd be cleaner to have these all just be different cases. Also, does a number always have to be positive?

djih replied:
In the case of SDK options yes, they need to be positive. The numerical options are all thresholds and limits such as max event count, session timeout, upload batch size, event upload threshold, event upload period in millis. These would likely break the SDK with values of 0 or negative.

paladin8 commented on src/amplitude.js:
```diff
@@ -433,33 +552,45 @@ Amplitude.prototype.setUserId = function(userId) {
+ * @example: amplitude.setOptOut(true);
+ */
+Amplitude.prototype.setOptOut = function setOptOut(enable) {
+  if (!this._apiKeySet('setOptOut()') || !utils.validateInput(enable, 'enable', 'boolean')) {
```
Have we always required API key to be set before any of these functions can be called, i.e. before multiple apps?

djih replied:
Actually I might need to rethink this. In v1 of multiple apps, each instance had its own cookie to save different config settings. The way we name the cookies is amplitude_ first 6 chars of the api key. For that reason, any public functions that logged an event, or interacted with the cookie in any way (modifying/saving/loading) would need the apiKey set. In v2 we actually namespace on the instance name instead of the apiKey, so it actually might not be necessary to always check for the apiKey now. Of course any of the user property stuff (like clearUserProperties) would still need the check since they trigger an identify event, which will post a request.

paladin8 commented on src/amplitude.js:
```diff
@@ -680,104 +842,110 @@ Amplitude.prototype.logRevenue = function(price, quantity, product) {
-    this._unsentEvents = filteredEvents;
+Amplitude.prototype.removeEvents = function removeEvents(maxEventId, maxIdentifyId) {
+  _removeEvents(this, '_unsentEvents', maxEventId);
+  _removeEvents(this, '_unsentIdentifys', maxIdentifyId);
```
Not sure why "this" needs to be passed in separately here as a scope. Can we just pass in ```this._unsentEvents``` and ```this._unsentIdentifys```?

djih replied:
In the helper function _removeEvents we actually create new arrays for the filtered events which overwrite the pointers in this._unsentEvents and this._unsentIdentifys (basically to support this line: `scope[eventQueue] = filteredEvents;`).

paladin8 commented on src/constants.js:
```diff
@@ -0,0 +1,19 @@
+module.exports = {
+  API_VERSION: 2,
+  MAX_STRING_LENGTH: 1024,
```
Let's increase this to 4096. In the future we may make the backend support larger strings, so we don't want the SDK to truncate them early.

djih replied:
hallelujah

paladin8 commented:
really solid changes, only worry is that it is pretty big so we just want to make sure we manually test enough and that upgrades work, etc. lgtm from a code perspective though!
